File: third/third.py
prior = {
    "a": 1,
    "z": 26,
    "A": 27,
    "Z": 52,
}


def getPrior(c):
    if "a" <= c <= "z":
        return ord(c) - ord("a") + 1
    elif "A" <= c <= "Z":
        return ord(c) - ord("A") + 27


# with open('input.txt') as f:
#     cnt = 0
#     for line in f:
#         line = line.strip()
#         firstpart, secondpart = line[:len(line)//2], line[len(line)//2:]
#         firstpart = set(firstpart)
#         secondpart = set(secondpart)
#         intersect = list(firstpart.intersection(secondpart))
#         for i in intersect:
#             cnt += getPrior(i)

def get_intersect(getPrior, three_lines, cnt):
    first, second, third = set(three_lines[0]), set(
        three_lines[1]), set(three_lines[2])
    intersect = list(first.intersection(second, third))
    cnt += getPrior(intersect[0])
    three_lines = []
    return three_lines, cnt


with open('input.txt') as f:
    three_lines = []
    cnt = 0
    for line in f:
        if len(three_lines) == 3:
            three_lines, cnt = get_intersect(getPrior, three_lines, cnt)
        three_lines.append(line.strip())
    if len(three_lines) == 3:
        three_lines, cnt = get_intersect(getPrior, three_lines, cnt)

print(cnt)

File: third/third.mjs
import { readFileSync } from 'node:fs';

const lines = readFileSync('input.txt', { encoding: 'utf-8' }) // read day??.txt content
  .replace(/\r/g, '') // remove all \r characters to avoid issues on Windows
  .trim() // Remove starting/ending whitespace
  .split('\n'); // Split on newline
// .map(Number); // Parse each line into a number

function calc_value(letter) {
  if (/[a-z]/.test(letter)) {
    return letter.charCodeAt(0) - 96;
  } else {
    return letter.charCodeAt(0) - 65 + 27;
  }
}

function part1() {
  //do something here
  const result = lines.map((line) => {
    const first = [...line.slice(0, Math.floor(line.length / 2))];
    const second = [...line.slice(Math.floor(line.length / 2))];

    const firstSet = new Set(first);
    const intersection = second.filter((value) => firstSet.has(value));
    const duplications = [...new Set(intersection)];

    return calc_value(duplications[0]);
  });
  console.log(result.reduce((a, b) => a + b, 0));
}

function part2() {
  //do something here
  let sum = 0;
  for (let i = 0; i < lines.length; i += 3) {
    const backpack = [[...lines[i]], [...lines[i + 1]], [...lines[i + 2]]];

    let set = new Set(backpack[0]);
    let intersection = backpack[1].filter((value) => set.has(value));

    set = new Set(intersection);
    intersection = backpack[2].filter((value) => set.has(value));

    const duplications = [...new Set(intersection)];
    sum += calc_value(duplications[0]);
  }

  console.log(sum);
}

part1();
part2();
